Fix `M0` click not consumed

leptun opened the pull request:
Fix the need to click the knob twice to dismiss `M0`. Also some code optimization.

Flash: -90B
SRAM: -2B

leptun committed:
Fix `M0` click not consumed

Fix the need to click the knob twice to dismiss `M0`.

Flash: -54B
SRAM: -2B

gudnimg commented:
Note if we merge, we can probably close https://github.com/prusa3d/Prusa-Firmware/pull/4077 :)

gudnimg committed:
Optimise M0/M1 code size

Replace two bool variables with one.

It's not obvious but (!hasP && !hasS) is equal to
!(hasP || hasS)

Note: expiration_time_set = hasP || hasS
Truth table:
|--------------------------------------------------|
| hasP| hasS| (!hasP && !hasS)|  !(hasP || hasS) |
|------|-----|--------------------|-----------------|
|   0    |  0   |               1            |            1           |
|   0    |  1   |               0            |            0           |
|   1    |  0   |               0            |            0           |
|   1    |  1   |               0            |            0           |
|--------------------------------------------------|

Change in memory:
Flash: -36 bytes
SRAM: 0 bytes

gudnimg reviewed (approved):
Using the changes in this PR fixes the issue (which is easily reproduced). Now only one knob click is needed :) And long-presses are ignored which is good.
